tests/testthat/helper-TealAppDriver.R

```diff
@@ -23,3 +23,41 @@ init_teal_app_driver <- function(...) {
     .package = "teal"
   )
 }
+
+ns_dataset <- function(prefix, suffix, dataset) {
+  sprintf("%s-dataset_%s_singleextract-%s", prefix, dataset, suffix)
+}
+
+test_plot_changes_no_errors <- function(app_driver, input_id, value, ws) {
+  test_object_no_changes(app_driver, input_id, value, ws, "get_active_module_pws_output")
+}
+
+test_table_changes_no_errors <- function(app_driver, input_id, value, ws) {
+  test_object_no_changes(app_driver, input_id, value, ws, "get_active_module_tws_output")
+}
+
+test_object_no_changes <- function(app_driver, input_id, value, ws, fun) {
+  object <- app_driver[[fun]]()(ws)
+  app_driver$set_active_module_input(input_id, value)
+  app_driver$expect_no_validation_error()
+  testthat::expect_false(identical(object, app_driver[[fun]]()(ws)))
+}
+
+test_validation_error <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL) {
+  app_driver$set_active_module_input(input_id, value)
+  app_driver$expect_validation_error()
+
+  element <- ifelse(!is.null(input), sprintf("%s .shiny-validation-message", input), validation)
+
+  if (!is.null(message)) {
+    testthat::expect_match(
+      app_driver$active_module_element_text(element),
+      message
+    )
+  }
+}
+
+test_no_validation_error <- function(app_driver, input_id, value) {
+  app_driver$set_active_module_input(input_id, value)
+  app_driver$expect_no_validation_error()
+}
```

Comment:
```suggestion
}

test_that_input_validation <- function(input_id, value = character(0L), validation_path, validation_message, test_desc) {
  testthat::test_that(
    desc = test_desc,
    code = {
      app_driver$set_active_module_input(input_id, value)
      app_driver$expect_validation_error()
      testthat::expect_match(app_driver$active_module_element_text(
        sprintf(
          "%s .shiny-validation-message", validation_path
        )),
        validation_message)
    }
  )
}
```

following from the discussion here https://github.com/insightsengineering/teal.modules.clinical/issues/1108#issuecomment-2078875272